### PR TITLE
 build: use manifest hooks for dev server proxy and fix hot reload for charts

### PR DESCRIPTION
**Pull request by ktmud**
### CATEGORY

- [x] Build / Development Environment

### SUMMARY

This updates how `wepback-dev-server` proxy reads asset manifest. Instead of reading from the generated `manifest.json` file, we use webpack hooks.

This is needed because sometimes proxy server will read `manifest.json` before it is fully updated, resulting the proxy to return the old code for an updated module on a hard page refresh.

Also added some docs for how to edit plugins as a followup to https://github.com/apache/incubator-superset/pull/9326 . 

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
N/A

### TEST PLAN

Before the change:

- Start the dev server, go to any page
- Edit a React component in that page, hot load should work, you component is updated
- Refresh the page, you component is served with the old code

After the change:
- You component should reflect the updated code.

### ADDITIONAL INFORMATION

N/A

### REVIEWERS

@rusackas @kristw

### CODE REVIEW

File: superset-frontend/webpack.config.js
Comment (ktmud) [inline superset-frontend/webpack.config.js]: Replaced [WebpackAssetsManafestPlugin](https://github.com/webdeveric/webpack-assets-manifest) with [WebpackManifestPlugin](https://github.com/danethurber/webpack-manifest-plugin) because the latter is recommended by Webpack's official docs and seem to have more users.

**Review by ktmud — COMMENTED** [review]
This one does not solve the hot reload issue with Charts I mentioned in #9326 , but does help by making sure the dev server always serve the latest code for updated modules.

File: CONTRIBUTING.md
Comment (rusackas) [inline CONTRIBUTING.md]: Same teensy tiny grammar nit from above :)
```suggestion
### Improving visualizations
```

File: CONTRIBUTING.md
Comment (rusackas) [inline CONTRIBUTING.md]: It's probably two levels back, so adjusting the sample path. Adding the means to link all packages will also help people
```suggestion
npm link ../../superset-ui-plugins/packages/superset-ui-[PLUGIN NAME]
# or to link all plugin packages:
# npm link ../../superset-ui-plugins/packages/*
```

File: CONTRIBUTING.md
Comment (rusackas) [inline CONTRIBUTING.md]:
```suggestion
Note that every time you do `npm install`, you will lose the symlink(s) and may have to run `npm link` again.
```

File: CONTRIBUTING.md
Comment (rusackas) [inline CONTRIBUTING.md]: Thought this might also be worth mentioning...
```suggestion

Note that when plugin packages are linked with `npm link` they will automatically load files from the plugin's `/src` directory.

```

**Review by rusackas — COMMENTED** [review]
Code looks good to me... I'll pull it and manually test it when time allows in the morning, before officially approving. I made a few minor edit suggestions to the documentation, which I hope help. Thanks for these improvements!

File: superset-frontend/webpack.config.js
Comment (ktmud) [inline superset-frontend/webpack.config.js]: Note that the new `manifest.json` will now not emit file entries at the root, e.g.:

```json
{
    "filename.js": "filename.[hash].js"
}
```

It will not only contain an `entrypoints` section (the only thing that's been used by both Flask and `webpack-dev-server`), a list of CSS/JS chunk files for each bundle.

**Review by rusackas — APPROVED** [review]
Welll... I spent some time today trying to replicate the issue according to the test plan. When I edit code of components (I was picking on dashboard tabs, mainly) I see the changes via hot reload. When I refresh, the changes are still there. 

That said, with the changes, everything still seems to work great, so I'm not seeing any regression either. 

I think the documentation in this PR is also important, and again the code looks good to me, so I'll approve this, but see if anyone else wants to weigh in on the bug reproduction issue before merging.

**Comment by ktmud** [comment]
> but see if anyone else wants to weigh in on the bug reproduction issue before merging.

To be specific, I saw the issue (updated code getting reverted to initial build) on Chart Explore page when I edited both the source file of the symlinked Big Number plugin and `src/chart/Chart.jsx`.

Also still seeing hot reload not working for symlinked components and was not able to fix it. Would be great if someone can help confirm whether this is reproducible, too.

**Comment by ktmud** [comment]
FINALLY managed to fix hot-reload issue on the Explore page. The problem is multifold.

1. `shouldComponentUpdate` is probably too strict. I [changed it](https://github.com/apache/incubator-superset/pull/9333/commits/7589b46bf127979b2dc24e518d94ecc5d4e52a6d#diff-b8d51b9ed2063863add2f7509be1b719R91) to only return false when `chartStatus` is updated from `success` to `rendered`.
2. Because `SuperChart` uses [memoized async loaders](https://github.com/apache-superset/superset-ui/blob/26ed9e31c6fd8950b01f8170836695cc9823192a/packages/superset-ui-chart/src/components/SuperChartCore.tsx#L93-L135) to load chart component, sometimes it confuses `react-hot-loader` when the module should be reloaded. The hack is to [add a unique key](https://github.com/apache/incubator-superset/pull/9333/commits/7589b46bf127979b2dc24e518d94ecc5d4e52a6d#diff-b8d51b9ed2063863add2f7509be1b719R233) to `SuperChart` component in `ChartRenderer` so it will always be rerendered.

Unfortunately, sometimes hot reload still doesn't work for charts in the dashboards. Considering the Explore page is normally where developers test + develop the charts, and they can always do a "cold" refresh, I'll stop digging and leave it as is. I did, however, [rewrite](https://github.com/apache/incubator-superset/pull/9333/commits/d5c8fd1caf482ea7ebd6f53bc00152cddb83fd99) `dashboard/App.jsx` and `dashboard/index.jsx` to avoid the same "<Provider> does not support change store on the fly" error as seen  [in the Explore page](https://github.com/apache/incubator-superset/pull/9326#discussion_r394777930).

File: superset-frontend/src/chart/ChartRenderer.jsx
Comment (ktmud) [inline superset-frontend/src/chart/ChartRenderer.jsx]: This does not change what users saw, but does avoid rendering the chart two times at hot reload.  See: https://github.com/gaearon/react-hot-loader#out-of-bound-warning

**Comment by rusackas** [comment]
Not sure why the PR page here shows Travis build is pending (seemingly forever) but if I click the "Details" link, it looks like it all worked great. Tested this locally, and happy to merge when the light turns green.

**Comment by ktmud** [comment]
@rusackas this CI is now green. Can you take another look?

**Comment by ktmud** [comment]
Hold on.. the updated `shouldComponentUpdate` logic is cause some performance issues in dashboards.

File: superset-frontend/src/chart/ChartRenderer.jsx
Comment (ktmud) [inline superset-frontend/src/chart/ChartRenderer.jsx]: I wrote my own depth-limited `deepEqual` function. Would be happy to use any existing library or util function if it exists.

File: superset-frontend/src/chart/ChartRenderer.jsx
Comment (ktmud) [inline superset-frontend/src/chart/ChartRenderer.jsx]: Not sure what we are using the `rendered` status for, but I'm keeping it here just in case.

**Review by ktmud — COMMENTED** [review]
Finally managed to fix hot reload on dashboards as well.

@kristw , can you also take a look at the latest commit, especially the changes to [ChartRenderer.jsx](https://github.com/apache/incubator-superset/commit/5c7b044cb0dcd28834cc8882575f3170fc1bb5c0#diff-b8d51b9ed2063863add2f7509be1b719)?

File: superset-frontend/src/dashboard/index.jsx
Comment (kristw) [inline superset-frontend/src/dashboard/index.jsx]: oh nice. so this will get rid of the hot loader warning about store, right?

File: superset-frontend/webpack.config.js
Comment (kristw) [reply superset-frontend/webpack.config.js]: sgtm

File: superset-frontend/webpack.config.js
Comment (kristw) [inline superset-frontend/webpack.config.js]: keep or remove?

File: superset-frontend/src/chart/ChartRenderer.jsx
Comment (kristw) [inline superset-frontend/src/chart/ChartRenderer.jsx]: remove

File: superset-frontend/src/chart/ChartRenderer.jsx
Comment (kristw) [inline superset-frontend/src/chart/ChartRenderer.jsx]:
```ts
const { chartStatus } = this.props;
```

File: superset-frontend/src/chart/ChartRenderer.jsx
Comment (kristw) [inline superset-frontend/src/chart/ChartRenderer.jsx]: add `{}`. Not sure why `eslint` is not warning.

File: superset-frontend/src/chart/Chart.jsx
Comment (kristw) [inline superset-frontend/src/chart/Chart.jsx]: Is this change intended? I didn't see any reference within this PR.

File: superset-frontend/src/chart/ChartRenderer.jsx
Comment (kristw) [reply superset-frontend/src/chart/ChartRenderer.jsx]: The logic seems to be changed here, since the original code did not check `this.props.chartStatus`.

File: superset-frontend/src/chart/ChartRenderer.jsx
Comment (kristw) [inline superset-frontend/src/chart/ChartRenderer.jsx]: Could we make this call in one of the setup file instead? It seems a bit odd to call this here.

**Review by kristw — COMMENTED** [review]
Thank you for keep improving the dev environment. I think most changes look good to me. 

I am a bit worry about the `shouldComponentUpdate` logic change in `ChartRenderer.jsx` as that will affect one of the largest surface in Superset. 

Wonder if that alone can be separated to another PR, so if something happens and we need to revert. The remaining good work in this PR which are pretty safe IMO should not be affected and can easily get merged as well.

File: superset-frontend/src/dashboard/index.jsx
Comment (ktmud) [reply superset-frontend/src/dashboard/index.jsx]: Yes.

File: superset-frontend/src/chart/Chart.jsx
Comment (ktmud) [reply superset-frontend/src/chart/Chart.jsx]: Yes. I was seeing `chartStackTrace` being `undefined` on the initial rendering, then immediately updated to `null`. Thought it makes sense to have a consistent value.

File: superset-frontend/src/chart/ChartRenderer.jsx
Comment (ktmud) [reply superset-frontend/src/chart/ChartRenderer.jsx]: I think both are acceptable. Would actually prefer no `{}` if the conditions are simpler.

File: superset-frontend/src/chart/ChartRenderer.jsx
Comment (ktmud) [reply superset-frontend/src/chart/ChartRenderer.jsx]: I noticed this config is only needed for charts and `isDevMode` is used somewhere down below as well, so it was just convenient to put it here. Will move it to `preamble.js`.

File: superset-frontend/src/dashboard/App.jsx
Comment (kristw) [inline superset-frontend/src/dashboard/App.jsx]: This seems to be unused import.

**Review by kristw — APPROVED** [review]

File: superset-frontend/src/dashboard/App.jsx
Comment (ktmud) [reply superset-frontend/src/dashboard/App.jsx]: Cleaned up!

**Pull request merged**
